# Project

## Load libraries and setup environment

In [1]:
# Import libraries
from datetime import timedelta
import os

import numpy as np
import pandas as pd
import re
import matplotlib.pyplot as plt
import matplotlib.dates as mdates

from IPython.display import display, HTML, Image
%matplotlib inline

plt.style.use('ggplot')
plt.rcParams.update({'font.size': 20})

# Access data using Google BigQuery.
from google.colab import auth
from google.cloud import bigquery

# authenticate
auth.authenticate_user()

# Set up environment variables
project_id = 'fast-oxide-395508'
if project_id == 'CHANGE-ME':
  raise ValueError('You must change project_id to your GCP project.')
os.environ["GOOGLE_CLOUD_PROJECT"] = project_id

# Read data from BigQuery into pandas dataframes.
def run_query(query, project_id=project_id):
  return pd.io.gbq.read_gbq(
      query,
      project_id=project_id,
      dialect='standard')

# set the dataset
# if you want to use the demo, change this to mimic_demo
dataset = 'mimiciv'

## Get dataset

### Arterial hypoxemia (arterial oxygen tension [Pa02]/fraction of inspired oxygen [FiO2] <300)  
Module hosp doesn't have the these data, so we get data from module icu

In [41]:
# Module icu:
query = """
SELECT *
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%inspired%" OR LOWER(label) LIKE "%arterial%pressure" OR LOWER(label) LIKE "%base%excess%" OR LOWER(label) LIKE "%urine%"
"""

df = run_query(query)
print(df)

    itemid                         label                  abbreviation  \
0   225454                 Urine Culture                 Urine Culture   
1   220224          Arterial O2 pressure                PO2 (Arterial)   
2   220235         Arterial CO2 Pressure               PCO2 (Arterial)   
3   220799     ZSpecific Gravity (urine)     ZSpecific gravity (urine)   
4   224828          Arterial Base Excess          Arterial Base Excess   
5   227471      Specific Gravity (urine)      Specific Gravity (urine)   
6   224015                  Urine Source                  Urine Source   
7   224016                   Urine Color                   Urine Color   
8   224876              Urine Appearance              Urine Appearance   
9   226566     Urine and GU Irrigant Out     Urine and GU Irrigant Out   
10  226627                      OR Urine                      OR Urine   
11  226631                    PACU Urine                    PACU Urine   
12  227489  GU Irrigant/Urine Volume O

We find "Arterial O2 pressure" and "Inspired O2 Fraction" is what we want

In [ ]:
query = f"""
SELECT
  d1.subject_id,
  d1.hadm_id,
  d1.stay_id,
  d1.itemid,
  d1.valuenum AS PaO2,
  d1.valueuom,
  d2.valuenum AS FiO2,
  (d1.valuenum / NULLIF(d2.valuenum, 0)) AS ratio
FROM
  (
    SELECT *
    FROM `physionet-data.mimiciv_icu.chartevents`
    WHERE itemid = 220224
  ) d1
JOIN
  (
    SELECT *
    FROM `physionet-data.mimiciv_icu.chartevents`
    WHERE itemid = 223835
  ) d2
ON
  d1.subject_id = d2.subject_id
AND
  d1.hadm_id = d2.hadm_id
AND
  d1.stay_id = d2.stay_id
WHERE
  (d1.valuenum / NULLIF(d2.valuenum, 0)) < 300
"""

df = run_query(query)
Arterial_hypoxemia_subject_id = df['subject_id'].unique()
Arterial_hypoxemia_subject_id = pd.DataFrame(Arterial_hypoxemia_subject_id, columns=['subject_id'])
Arterial_hypoxemia_subject_id

subject_id
0        12627553
1        15199651
2        13632970
3        15620117
4        12889874
...           ...
24309    19827412
24310    10267064
24311    12538508
24312    16605907
24313    13921082

[24314 rows x 1 columns]

In [46]:
subject_ids = patients_cohort['subject_id'].tolist()

# 先从outputevents中获取相关数据
query = f"""
SELECT *
FROM `physionet-data.mimiciv_icu.outputevents`
WHERE itemid = 226627 AND subject_id IN ({','.join(map(str, subject_ids))})
"""

df = run_query(query)

# 在Python中进行合并和筛选
df = df.merge(patients_cohort[['subject_id', 'weaning_start_time']], on="subject_id", how="inner")

# 筛选出weaning_start_time前24小时的数据
mask = (df['charttime'] > df['weaning_start_time'] - pd.Timedelta(hours=24)) & (df['charttime'] <= df['weaning_start_time'])
df_filtered = df[mask]
df_filtered

subject_id   hadm_id   stay_id  caregiver_id           charttime  \
0       10004422  21255400  32155744         87605 2111-01-17 19:11:00   
1       10004457  23251352  31494479         53913 2141-12-17 15:00:00   
2       10009628  25926192  35258379         86518 2153-09-19 13:40:00   
5       10014729  28889419  33558396         53913 2125-02-27 14:45:00   
6       10017437  28440971  33439506         12098 2118-04-17 15:03:00   
...          ...       ...       ...           ...                 ...   
3476    19968039  20687784  35528180         91879 2132-04-20 23:21:00   
3477    19972172  23625602  39366389         49403 2156-10-20 22:40:00   
3478    19975044  20044208  37790066         28029 2155-02-24 15:51:00   
3480    19984052  28687121  33735777         28791 2184-11-12 13:00:00   
3481    19993726  20161163  31439824         89212 2166-09-14 17:41:00   

               storetime  itemid   value valueuom  weaning_start_time  
0    2111-01-17 19:11:00  226627   300.0       ml 2111-01-18 06:31:00  
1    2141-12-17 17:09:00  226627   675.0       ml 2141-12-17 18:04:00  
2    2153-09-19 14:08:00  226627   720.0       ml 2153-09-19 18:18:00  
5    2125-02-27 17:38:00  226627   415.0       ml 2125-02-27 19:05:00  
6    2118-04-17 15:03:00  226627   550.0       ml 2118-04-17 21:00:00  
...                  ...     ...     ...      ...                 ...  
3476 2132-04-20 23:21:00  226627  1000.0       ml 2132-04-21 10:45:00  
3477 2156-10-21 03:06:00  226627   100.0       ml 2156-10-21 14:39:00  
3478 2155-02-24 15:51:00  226627  1250.0       ml 2155-02-24 21:47:00  
3480 2184-11-12 15:50:00  226627   670.0       ml 2184-11-12 17:07:00  
3481 2166-09-14 17:42:00  226627   425.0       ml 2166-09-15 14:04:00  

[1958 rows x 10 columns]

### Arterial hypotension (eg, systolic blood pressure (SBP] <90 mmHg, mean arterial pressure (MAP) <70mmHg)

In [ ]:
query = f"""
SELECT label, itemid
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%systolic%"
"""

df = run_query(query)
df

label  itemid
0              Aortic Pressure Signal - Systolic  228152
1   Pulmonary Atrtery Pressure Signal - Systolic  229669
2    Left Ventricular Pressure Signal - Systolic  229899
3             Pulmonary Artery Pressure systolic  220059
4                  RV systolic pressure(PA Line)  226850
5                  PA systolic pressure(PA Line)  226852
6               Arterial Blood Pressure systolic  220050
7           Non Invasive Blood Pressure systolic  220179
8            Manual Blood Pressure Systolic Left  224167
9                                ART BP Systolic  225309
10          Manual Blood Pressure Systolic Right  227243

We choose "Non Invasive Blood Pressure systolic"

In [ ]:
query = f"""
SELECT label, itemid
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%arterial%mean%" OR LOWER(label) LIKE "%oxygenation%index%"
"""

df = run_query(query)
df

label  itemid
0  Arterial Blood Pressure mean  220052

We choose "Arterial Blood Pressure mean"

In [ ]:
query = f"""
SELECT
  t1.subject_id,
  t1.hadm_id,
  t1.stay_id,
  t1.itemid,
  t1.valuenum AS SBP,
  t1.valueuom,
  t2.itemid,
  t2.valuenum AS MAP,
  t2.valueuom
FROM
  (
    SELECT *
    FROM `physionet-data.mimiciv_icu.chartevents`
    WHERE itemid = 220179 AND valuenum < 90
  ) t1
JOIN
  (
    SELECT *
    FROM `physionet-data.mimiciv_icu.chartevents`
    WHERE itemid = 220052 AND valuenum < 70
  ) t2
ON
  t1.subject_id = t2.subject_id
AND
  t1.hadm_id = t2.hadm_id
AND
  t1.stay_id = t2.stay_id
"""

df = run_query(query)
Arterial_hypotension_subject_id = df['subject_id'].unique()
Arterial_hypotension_subject_id = pd.DataFrame(Arterial_hypotension_subject_id, columns=['subject_id'])
Arterial_hypotension_subject_id

subject_id
0        13612783
1        18756985
2        15112972
3        17030962
4        19097438
...           ...
11258    13783676
11259    13730292
11260    19889178
11261    18994071
11262    16160074

[11263 rows x 1 columns]

### Leukocytosis (white blood cell (WBC] count >12,000 microL-1) or leukopenia (WBC count <4000microL-1)

In [60]:
query = f"""
SELECT label, itemid
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%wbc%" OR LOWER(label) LIKE "%anion%"
"""

df = run_query(query)
df

label  itemid
0                WBC  220546
1          Anion gap  227073
2   WbcApacheIIScore  226779
3   WBCApacheIIValue  226780
4       WBC_ApacheIV  227062
5  WBCScore_ApacheIV  227063

We choose "WBC"

In [ ]:
query = f"""
SELECT *
FROM `physionet-data.mimiciv_icu.chartevents`
WHERE itemid = 220546 AND (valuenum > 12000 OR valuenum < 4000)
"""

df = run_query(query)
Leukocytosis_subject_id = df['subject_id'].unique()
Leukocytosis_subject_id = pd.DataFrame(Leukocytosis_subject_id, columns=['subject_id'])
Leukocytosis_subject_id

subject_id
0        16433404
1        17589125
2        18680755
3        19650110
4        10437898
...           ...
49475    11494577
49476    10783607
49477    19946367
49478    17888905
49479    12934260

[49480 rows x 1 columns]

### Thrombocytopenia (platelet count <100,000 microL-1)

In [ ]:
query = f"""
SELECT *
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%platelet%" or LOWER(label) LIKE "%creatinine%"
"""

df = run_query(query)
df

itemid                     label              abbreviation      linksto  \
0  220615        Creatinine (serum)        Creatinine (serum)  chartevents   
1  225678           ZPlatelet Count           ZPlatelet Count  chartevents   
2  229761  Creatinine (whole blood)  Creatinine (whole blood)  chartevents   
3  227457            Platelet Count            Platelet Count  chartevents   
4  226751   CreatinineApacheIIScore   CreatinineApacheIIScore  chartevents   
5  226752   CreatinineApacheIIValue   CreatinineApacheIIValue  chartevents   
6  227005       Creatinine_ApacheIV       Creatinine_ApacheIV  chartevents   
7  225170                 Platelets                 Platelets  inputevents   
8  226369        OR Platelet Intake        OR Platelet Intake  inputevents   
9  227071      PACU Platelet Intake      PACU Platelet Intake  inputevents   

                  category unitname        param_type  lownormalvalue  \
0                     Labs     None           Numeric             NaN   
1                     Labs     None           Numeric             NaN   
2                     Labs     None           Numeric             NaN   
3                     Labs     None  Numeric with tag             NaN   
4       Scores - APACHE II     None           Numeric             NaN   
5       Scores - APACHE II     None           Numeric             NaN   
6   Scores - APACHE IV (2)     None           Numeric             NaN   
7  Blood Products/Colloids       mL          Solution             NaN   
8  Blood Products/Colloids       mL          Solution             NaN   
9  Blood Products/Colloids       mL          Solution             NaN   

   highnormalvalue  
0              NaN  
1              NaN  
2              NaN  
3              NaN  
4              NaN  
5              NaN  
6              NaN  
7              NaN  
8              NaN  
9              NaN

We choose "Platelet Count"

In [ ]:
query = f"""
SELECT *
FROM `physionet-data.mimiciv_icu.chartevents`
WHERE itemid = 227457 AND valuenum < 100000
"""

df = run_query(query)
Thrombocytopenia_subject_id = df['subject_id'].unique()
Thrombocytopenia_subject_id = pd.DataFrame(Thrombocytopenia_subject_id, columns=['subject_id'])
Thrombocytopenia_subject_id

subject_id
0        17914766
1        18259787
2        16178416
3        17924051
4        18151060
...           ...
49485    19157537
49486    17795151
49487    12493025
49488    10641032
49489    12618032

[49490 rows x 1 columns]

### Heart rate >90 beats/min

In [ ]:
query = f"""
SELECT label, itemid
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%heart%rate%" OR LOWER(label) LIKE "%respiratory%rate%"
"""

df = run_query(query)
df

label  itemid
0         Heart rate Alarm - High  220046
1          Heart Rate Alarm - Low  220047
2                Respiratory Rate  220210
3          Respiratory Rate (Set)  224688
4  Respiratory Rate (spontaneous)  224689
5        Respiratory Rate (Total)  224690
6                      Heart Rate  220045

We choose "Heart Rate"

In [ ]:
query = f"""
SELECT *
FROM `physionet-data.mimiciv_icu.chartevents`
WHERE itemid = 220045 AND valuenum > 90
"""

df = run_query(query)
Heart_rate_subject_id = df['subject_id'].unique()
Heart_rate_subject_id = pd.DataFrame(Heart_rate_subject_id, columns=['subject_id'])
Heart_rate_subject_id

subject_id
0        10390732
1        16748272
2        10440899
3        17167982
4        18595258
...           ...
41622    15591535
41623    10911585
41624    16774594
41625    15181772
41626    15854999

[41627 rows x 1 columns]

### patients who have invasive Ventilation

In [ ]:
query = f"""
SELECT char.*, dlab.label
FROM `physionet-data.mimiciv_icu.procedureevents` AS char
INNER JOIN `physionet-data.mimiciv_icu.d_items` AS dlab
 ON char.itemid = dlab.itemid
WHERE char.itemid = 225792
ORDER BY subject_id
"""

df_invasive_ventilation = run_query(query)
invasive_ventilation_subject_id = df_invasive_ventilation['subject_id'].unique()
invasive_ventilation_subject_id = pd.DataFrame(invasive_ventilation_subject_id, columns=['subject_id'])
df_invasive_ventilation

subject_id   hadm_id   stay_id  caregiver_id           starttime  \
0        10001884  26184834  37510196         42150 2131-01-11 04:40:00   
1        10001884  26184834  37510196         91332 2131-01-13 04:00:00   
2        10001884  26184834  37510196         91332 2131-01-15 04:07:00   
3        10002013  23581541  39060235         27479 2160-05-18 14:19:00   
4        10002428  23473524  35479615         27479 2156-05-11 16:05:00   
...           ...       ...       ...           ...                 ...   
30705    19999068  21606769  30143796         29141 2161-08-25 15:35:00   
30706    19999287  20175828  35165301         15882 2197-08-06 19:29:00   
30707    19999442  26785317  32336619         86640 2148-11-19 19:00:00   
30708    19999840  21033226  38978960          <NA> 2164-09-12 09:41:00   
30709    19999987  23865745  36195440         31252 2145-11-02 23:00:00   

                  endtime               storetime  itemid    value valueuom  \
0     2131-01-12 17:40:00 2131-01-12 17:49:00.000  225792   2220.0      min   
1     2131-01-19 17:45:00 2131-01-19 18:44:00.000  225792   9465.0      min   
2     2131-01-19 17:43:00 2131-01-19 18:44:00.000  225792   6576.0      min   
3     2160-05-18 18:01:00 2160-05-18 18:39:00.000  225792    222.0      min   
4     2156-05-20 10:45:00 2156-05-20 10:51:00.000  225792  12640.0      min   
...                   ...                     ...     ...      ...      ...   
30705 2161-08-28 13:36:00 2161-08-28 13:36:00.000  225792   4201.0      min   
30706 2197-08-07 09:20:00 2197-08-07 09:23:00.000  225792    831.0      min   
30707 2148-11-20 14:15:00 2148-11-20 14:17:00.000  225792   1155.0      min   
30708 2164-09-17 16:35:00 2164-09-17 16:35:22.453  225792   7614.0      min   
30709 2145-11-03 20:42:00 2145-11-03 21:07:00.000  225792   1302.0      min   

       ... linkorderid ordercategoryname  ordercategorydescription  \
0      ...     3830120       Ventilation         ContinuousProcess   
1      ...     4465887       Ventilation         ContinuousProcess   
2      ...     1861924       Ventilation         ContinuousProcess   
3      ...     4169380       Ventilation         ContinuousProcess   
4      ...     3976442       Ventilation         ContinuousProcess   
...    ...         ...               ...                       ...   
30705  ...      858585       Ventilation         ContinuousProcess   
30706  ...     9514927       Ventilation         ContinuousProcess   
30707  ...     9582145       Ventilation         ContinuousProcess   
30708  ...     7234916       Ventilation         ContinuousProcess   
30709  ...     9006313       Ventilation         ContinuousProcess   

       patientweight isopenbag continueinnextdept  statusdescription  \
0               65.0         1                  0    FinishedRunning   
1               65.0         1                  0    FinishedRunning   
2               65.0         1                  0    FinishedRunning   
3               96.0         1                  0    FinishedRunning   
4               48.4         1                  0    FinishedRunning   
...              ...       ...                ...                ...   
30705           55.8         1                  0    FinishedRunning   
30706           86.2         1                  0    FinishedRunning   
30707          107.5         1                  0    FinishedRunning   
30708           77.5         1                  0    FinishedRunning   
30709           60.0         1                  0    FinishedRunning   

       ORIGINALAMOUNT  ORIGINALRATE                 label  
0              2220.0           1.0  Invasive Ventilation  
1              9465.0           1.0  Invasive Ventilation  
2              6576.0           1.0  Invasive Ventilation  
3               222.0           1.0  Invasive Ventilation  
4             12640.0           1.0  Invasive Ventilation  
...               ...           ...                   ...  
30705          4201.0          

### merge all the above filters to get the final cohort of Sepsis patients

In [ ]:
merged_df = Arterial_hypotension_subject_id.merge(invasive_ventilation_subject_id, on='subject_id', how='inner')
merged_df = merged_df.merge(Leukocytosis_subject_id, on='subject_id', how='inner')
merged_df = merged_df.merge(Thrombocytopenia_subject_id, on='subject_id', how='inner')
#merged_df = merged_df.merge(Heart_rate_subject_id, on='subject_id', how='inner')
sepsis_subject_ids = merged_df['subject_id'].unique()

print(sepsis_subject_ids)

<IntegerArray>
[13612783, 18756985, 15112972, 17030962, 19097438, 10003400, 19461458,
 18346662, 19420664, 10813665,
 ...
 17355264, 11152474, 11045637, 12882485, 17809884, 13783676, 13730292,
 19889178, 18994071, 16160074]
Length: 11247, dtype: Int64


exclude patients under 18 and repeated ICU admission

In [ ]:
query = """
SELECT subject_id
FROM `physionet-data.mimiciv_hosp.patients`
WHERE anchor_age >= 18
"""

age_filtered_df = run_query(query)

final_df = merged_df.merge(age_filtered_df, on='subject_id', how='inner')
final_df

subject_id
0       10098993
1       17559288
2       16467047
3       11750559
4       18454868
...          ...
9434    11230804
9435    18482037
9436    15276613
9437    17411386
9438    12597051

[9439 rows x 1 columns]

In [ ]:
query = """
SELECT subject_id
FROM `physionet-data.mimiciv_icu.icustays`
GROUP BY subject_id
HAVING COUNT(*) > 1
"""

multiple_admission_subjects = run_query(query)

final_df_2 = final_df[~final_df['subject_id'].isin(multiple_admission_subjects['subject_id'])]
final_df_2

subject_id
2       16467047
5       17087880
6       19556965
9       10222191
10      10252795
...          ...
9430    11102902
9431    14477446
9434    11230804
9436    15276613
9437    17411386

[5521 rows x 1 columns]

In [ ]:
final_df_2.to_csv('filtered_patients_final.csv', index=False)

### weaning success

#### No intubation or invasive ventilation within 48 h after weaning, and noninvasive ventilation time was shorter than 48 h after weaning

In [ ]:
query = f"""
SELECT *
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%wean%"
"""

df = run_query(query)
df

itemid label abbreviation      linksto     category unitname param_type  \
0  223870  Wean         Wean  chartevents  Respiratory     None       Text   

   lownormalvalue  highnormalvalue  
0             NaN              NaN

In [ ]:
query = f"""
SELECT char.*
FROM `physionet-data.mimiciv_icu.chartevents` AS char
WHERE char.itemid = 223870
ORDER BY subject_id
"""

df_weaning = run_query(query)
df_weaning

Empty DataFrame
Columns: [subject_id, hadm_id, stay_id, caregiver_id, charttime, storetime, itemid, value, valuenum, valueuom, warning]
Index: []

用"Wean"找不到数据

In [ ]:
query = f"""
SELECT label, itemid, linksto
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%extubation%"
"""

df = run_query(query)
df

label  itemid          linksto
0      Unplanned Extubation (patient-initiated)  225468  procedureevents
1  Unplanned Extubation (non-patient initiated)  225477  procedureevents
2                                    Extubation  227194  procedureevents

In [ ]:
query = f"""
SELECT label, itemid, linksto
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%invasive ventilation%" OR LOWER(label) LIKE "%intubation%"
"""

df = run_query(query)
df

label  itemid          linksto
0                                          Intubation  224385  procedureevents
1                                Invasive Ventilation  225792  procedureevents
2                            Non-invasive Ventilation  225794  procedureevents
3              Mallampati classification (Intubation)  222871      chartevents
4                  Induction/Drugs/Doses (Intubation)  225271      chartevents
5                             Indication (Intubation)  225291      chartevents
6                          Mouth Opening (Intubation)  225294      chartevents
7                   Thyromental Distance (Intubation)  225295      chartevents
8                  Mandibular Prognatism (Intubation)  225296      chartevents
9                        Teeth/Dentition (Intubation)  225297      chartevents
10                              Neck ROM (Intubation)  225298      chartevents
11                      Mask Ventilation (Intubation)  225303      chartevents
12                     Blade / Equipment (Intubation)  225304      chartevents
13                    Number of Attempts (Intubation)  225306      chartevents
14                             ETT Depth (Intubation)  225585      chartevents
15                 Tracheal Confirmation (Intubation)  225590      chartevents
16                              ETT Size (Intubation)  225592      chartevents
17             Best Larynx View Obtained (Intubation)  225593      chartevents
18                  Timeout Performed by (Intubation)  226188      chartevents
19  Patient identified correctly by 2 means (Intub...  226189      chartevents
20          Barrier precautions in place (Intubation)  226191      chartevents
21                              ETT Type (Intubation)  226429      chartevents
22                         Complications (Intubation)  226430      chartevents
23                       Verification of (Intubation)  226431      chartevents
24                                 Drugs (Intubation)  226432      chartevents
25     Hand Cleansing prior to procedure (Intubation)  228065      chartevents
26  Unable to complete Airway Assessment (Intubation)  228066      chartevents
27                              Stylette (Intubation)  228067      chartevents
28                                Bougie (Intubation)  228068      chartevents
29                          Tube Secured (Intubation)  228069      chartevents
30          Barrier precautions in place (Intubation)  228643      chartevents
31                           Code Status (Intubation)  229784      chartevents
32                        Preoxygenation (Intubation)  229786      chartevents
33                        Rigid Stylette (Intubation)  229787      chartevents
34                              Propofol (Intubation)  227210      chartevents
35                              Ketamine (Intubation)  227211      chartevents
36                             Etomidate (Intubation)  227212      chartevents
37                            Vecuronium (Intubation)  227213      chartevents
38                        Cis-atracurium (Intubation)  227214      chartevents
39                       Succynlocholine (Intubation)  227215      chartevents
40                            Rocuronium (Intubation)  229788      chartevents
41                         Phenylephrine (Intubation)  229789      chartevents
42  Unable to complete Airway Assessment (Intubation)  225292      chartevents
43                 No Drugs Administered (Intubation)  225300      chartevents
44                                   RSI (Intubation)  225301      chartevents
45                       Pre-Oxygentated (Intubation)  225302      chartevents
46                              Stylette (Intubation)  225586      chartevents
47                                Bougie (Intubation)  225587      chartevents
48                          Tube Secured (Intubation)  225588      chartevents
49                      No Complications (Intubation)  226166      chartevents
50     Hand Cleansi

We choose "Extubation", "Intubation" and "Invasive Ventilation"

Extubation

In [ ]:
query = f"""
SELECT char.*, dlab.label
FROM `physionet-data.mimiciv_icu.procedureevents` AS char
INNER JOIN `physionet-data.mimiciv_icu.d_items` AS dlab
 ON char.itemid = dlab.itemid
WHERE char.itemid = 227194
ORDER BY subject_id
"""

df_weaning = run_query(query)
df_weaning

subject_id   hadm_id   stay_id  caregiver_id           starttime  \
0        10001884  26184834  37510196         35966 2131-01-12 17:40:00   
1        10002013  23581541  39060235         65197 2160-05-18 18:00:00   
2        10002428  23473524  35479615         42665 2156-05-20 10:45:00   
3        10002428  28662225  38875437          3009 2156-04-22 17:10:00   
4        10002760  28094813  31831386         50512 2141-04-21 08:00:00   
...           ...       ...       ...           ...                 ...   
22331    19998843  24842066  30988867         93405 2187-02-08 17:09:00   
22332    19999068  21606769  30143796         99850 2161-08-28 13:35:00   
22333    19999287  20175828  35165301         50512 2197-08-07 09:20:00   
22334    19999442  26785317  32336619         26402 2148-11-20 14:15:00   
22335    19999987  23865745  36195440         18014 2145-11-03 20:40:00   

                  endtime           storetime  itemid  value valueuom  ...  \
0     2131-01-12 17:41:00 2131-01-12 17:50:00  227194    1.0     None  ...   
1     2160-05-18 18:01:00 2160-05-18 17:59:00  227194    1.0     None  ...   
2     2156-05-20 10:46:00 2156-05-20 10:51:00  227194    1.0     None  ...   
3     2156-04-22 17:11:00 2156-04-22 17:10:00  227194    1.0     None  ...   
4     2141-04-21 08:01:00 2141-04-21 08:04:00  227194    1.0     None  ...   
...                   ...                 ...     ...    ...      ...  ...   
22331 2187-02-08 17:10:00 2187-02-08 17:15:00  227194    1.0     None  ...   
22332 2161-08-28 13:36:00 2161-08-28 13:37:00  227194    1.0     None  ...   
22333 2197-08-07 09:21:00 2197-08-07 09:23:00  227194    1.0     None  ...   
22334 2148-11-20 14:16:00 2148-11-20 14:17:00  227194    1.0     None  ...   
22335 2145-11-03 20:41:00 2145-11-03 20:40:00  227194    1.0     None  ...   

      linkorderid      ordercategoryname  ordercategorydescription  \
0         6470885  Intubation/Extubation                      Task   
1         3546453  Intubation/Extubation                      Task   
2         9794232  Intubation/Extubation                      Task   
3         9954431  Intubation/Extubation                      Task   
4         2857656  Intubation/Extubation                      Task   
...           ...                    ...                       ...   
22331     9357731  Intubation/Extubation                      Task   
22332      489079  Intubation/Extubation                      Task   
22333     4142066  Intubation/Extubation                      Task   
22334     8029280  Intubation/Extubation                      Task   
22335     7230124  Intubation/Extubation                      Task   

       patientweight isopenbag continueinnextdept  statusdescription  \
0               65.0         0                  0    FinishedRunning   
1               96.0         0                  0    FinishedRunning   
2               48.4         0                  0    FinishedRunning   
3               43.0         0                  0    FinishedRunning   
4               91.0         0                  0    FinishedRunning   
...              ...       ...                ...                ...   
22331           90.0         0                  0    FinishedRunning   
22332           55.8         0                  0    FinishedRunning   
22333           86.2         0                  0    FinishedRunning   
22334          107.5         0                  0    FinishedRunning   
22335           94.0         0                  0    FinishedRunning   

       ORIGINALAMOUNT  ORIGINALRATE       label  
0                 1.0           0.0  Extubation  
1                 1.0           0.0  Extubation  
2                 1.0           0.0  Extubation  
3                 1.0           0.0  Extubation  
4                 1.0           0.0  Extubation  
...               ...           ...         ...  
22331             1.0           0.0  Extubation  
22332             1.0           0.0  Extubation  
22333        

Intubation

In [ ]:
query = f"""
SELECT char.*, dlab.label
FROM `physionet-data.mimiciv_icu.procedureevents` AS char
INNER JOIN `physionet-data.mimiciv_icu.d_items` AS dlab
 ON char.itemid = dlab.itemid
WHERE char.itemid = 224385
ORDER BY subject_id
"""

df_intubation = run_query(query)
df_intubation

subject_id   hadm_id   stay_id  caregiver_id           starttime  \
0       10001884  26184834  37510196         42150 2131-01-11 04:30:00   
1       10003400  23559586  38383343         36212 2137-08-17 21:21:00   
2       10004401  27939719  31202136         58760 2144-04-11 05:03:00   
3       10004401  29988601  32773003         38908 2144-01-27 19:00:00   
4       10004401  29988601  32773003         42665 2144-01-30 13:00:00   
...          ...       ...       ...           ...                 ...   
8483    19997367  20617667  35616526         10976 2126-04-26 20:14:00   
8484    19997367  20617667  35616526         27479 2126-05-05 18:20:00   
8485    19999068  21606769  30143796         29141 2161-08-25 15:34:00   
8486    19999442  26785317  32336619         59028 2148-11-19 19:00:00   
8487    19999840  21033226  38978960         82641 2164-09-12 09:40:00   

                 endtime           storetime  itemid  value valueuom  ...  \
0    2131-01-11 04:31:00 2131-01-11 05:40:00  224385    1.0     None  ...   
1    2137-08-17 21:22:00 2137-08-17 21:21:00  224385    1.0     None  ...   
2    2144-04-11 05:04:00 2144-04-11 05:03:00  224385    1.0     None  ...   
3    2144-01-27 19:01:00 2144-01-27 19:12:00  224385    1.0     None  ...   
4    2144-01-30 13:01:00 2144-01-30 13:05:00  224385    1.0     None  ...   
...                  ...                 ...     ...    ...      ...  ...   
8483 2126-04-26 20:15:00 2126-04-26 21:14:00  224385    1.0     None  ...   
8484 2126-05-05 18:21:00 2126-05-05 18:20:00  224385    1.0     None  ...   
8485 2161-08-25 15:35:00 2161-08-25 15:45:00  224385    1.0     None  ...   
8486 2148-11-19 19:01:00 2148-11-19 21:31:00  224385    1.0     None  ...   
8487 2164-09-12 09:41:00 2164-09-12 09:41:00  224385    1.0     None  ...   

     linkorderid      ordercategoryname  ordercategorydescription  \
0        2338046  Intubation/Extubation                      Task   
1        9156137  Intubation/Extubation                      Task   
2         991564  Intubation/Extubation                      Task   
3        4299790  Intubation/Extubation                      Task   
4        6177150  Intubation/Extubation                      Task   
...          ...                    ...                       ...   
8483     5095062  Intubation/Extubation                      Task   
8484     8462542  Intubation/Extubation                      Task   
8485     6823223  Intubation/Extubation                      Task   
8486     3602635  Intubation/Extubation                      Task   
8487     5963844  Intubation/Extubation                      Task   

      patientweight isopenbag continueinnextdept  statusdescription  \
0              65.0         0                  0    FinishedRunning   
1              90.5         0                  0    FinishedRunning   
2             120.0         0                  0    FinishedRunning   
3              76.0         0                  0    FinishedRunning   
4              76.0         0                  0    FinishedRunning   
...             ...       ...                ...                ...   
8483           59.0         0                  0    FinishedRunning   
8484           59.0         0                  0    FinishedRunning   
8485           55.8         0                  0    FinishedRunning   
8486          107.5         0                  0    FinishedRunning   
8487           77.5         0                  0    FinishedRunning   

      ORIGINALAMOUNT  ORIGINALRATE       label  
0                1.0           0.0  Intubation  
1                1.0           0.0  Intubation  
2                1.0           0.0  Intubation  
3                1.0           0.0  Intubation  
4                1.0           0.0  Intubation  
...              ...           ...         ...  
8483             1.0           0.0  Intubation  
8484             1.0           0.0  Intubation  
8485             1.0           0.0  Intubation  
8486             1.0

noninvasive ventilation

In [ ]:
query = f"""
SELECT label, itemid, linksto
FROM `physionet-data.mimiciv_icu.d_items`
WHERE LOWER(label) LIKE "%non-invasive ventilation%"
"""

df = run_query(query)
df

label  itemid          linksto
0  Non-invasive Ventilation  225794  procedureevents

We choose "Non-Invasive Ventilation"

In [ ]:
query = f"""
SELECT char.*, dlab.label
FROM `physionet-data.mimiciv_icu.procedureevents` AS char
INNER JOIN `physionet-data.mimiciv_icu.d_items` AS dlab
 ON char.itemid = dlab.itemid
WHERE char.itemid = 225794
ORDER BY subject_id
"""

df_non_invasive_ventilation = run_query(query)
df_non_invasive_ventilation

subject_id   hadm_id   stay_id  caregiver_id           starttime  \
0       10000980  26913865  39765666          <NA> 2189-06-27 09:01:00   
1       10001884  26184834  37510196         31763 2131-01-12 21:30:00   
2       10002155  20345487  32358465         47007 2131-03-10 00:15:00   
3       10002428  20321825  34807493         99293 2156-04-30 22:54:00   
4       10002495  24982426  36753294          6579 2141-05-23 20:15:00   
...          ...       ...       ...           ...                 ...   
3006    19997293  20208898  38350444         58760 2123-10-20 02:00:00   
3007    19998330  24096838  33428243         80339 2178-11-28 05:00:00   
3008    19998330  24096838  33428243          <NA> 2178-11-29 06:13:00   
3009    19999287  20175828  35165301         97774 2197-08-05 05:22:00   
3010    19999987  23865745  36195440         10712 2145-11-03 22:00:00   

                 endtime               storetime  itemid   value valueuom  \
0    2189-06-27 20:38:00 2189-06-27 20:38:29.047  225794   697.0      min   
1    2131-01-13 04:00:00 2131-01-15 04:07:00.000  225794   390.0      min   
2    2131-03-10 07:59:00 2131-03-10 16:26:00.000  225794   464.0      min   
3    2156-05-02 05:03:00 2156-05-02 09:42:00.000  225794  1809.0      min   
4    2141-05-24 01:36:00 2141-05-24 06:47:00.000  225794   321.0      min   
...                  ...                     ...     ...     ...      ...   
3006 2123-10-20 04:03:00 2123-10-20 04:39:00.000  225794   123.0      min   
3007 2178-11-28 09:00:00 2178-11-29 06:14:00.000  225794   240.0      min   
3008 2178-11-29 21:29:00 2178-11-29 21:29:41.257  225794   916.0      min   
3009 2197-08-06 17:32:00 2197-08-06 19:29:00.000  225794  2170.0      min   
3010 2145-11-04 02:00:00 2145-11-04 18:07:00.000  225794   240.0      min   

      ... linkorderid ordercategoryname  ordercategorydescription  \
0     ...     5989583       Ventilation         ContinuousProcess   
1     ...     4809276       Ventilation         ContinuousProcess   
2     ...     1744937       Ventilation         ContinuousProcess   
3     ...     8673415       Ventilation         ContinuousProcess   
4     ...      183864       Ventilation         ContinuousProcess   
...   ...         ...               ...                       ...   
3006  ...     6650730       Ventilation         ContinuousProcess   
3007  ...      875619       Ventilation         ContinuousProcess   
3008  ...      228938       Ventilation         ContinuousProcess   
3009  ...     1972780       Ventilation         ContinuousProcess   
3010  ...     5191564       Ventilation         ContinuousProcess   

      patientweight isopenbag continueinnextdept  statusdescription  \
0              76.2         1                  0    FinishedRunning   
1              65.0         1                  0    FinishedRunning   
2              21.1         1                  0    FinishedRunning   
3              55.0         1                  0    FinishedRunning   
4              64.1         1                  0    FinishedRunning   
...             ...       ...                ...                ...   
3006          134.5         1                  0    FinishedRunning   
3007           83.0         1                  0    FinishedRunning   
3008           83.0         1                  0    FinishedRunning   
3009           86.2         1                  0    FinishedRunning   
3010           94.0         1                  0    FinishedRunning   

      ORIGINALAMOUNT  ORIGINALRATE                     label  
0              697.0           1.0  Non-invasive Ventilation  
1              390.0           1.0  Non-invasive Ventilation  
2              464.0           1.0  Non-invasive Ventilation  
3             1809.0           1.0  Non-invasive Ventilation  
4              321.0           1.0  Non-invasive Ventilation  
...              ...           ...                       ...  
3006           123.0           1.0  Non-invasive Ventilation  
3007  

In [ ]:
# Convert timestamp columns to datetime objects
df_weaning['endtime'] = pd.to_datetime(df_weaning['endtime'])
df_intubation['starttime'] = pd.to_datetime(df_intubation['starttime'])
df_invasive_ventilation['starttime'] = pd.to_datetime(df_invasive_ventilation['starttime'])
df_non_invasive_ventilation['starttime'] = pd.to_datetime(df_non_invasive_ventilation['starttime'])

# Find patients who underwent weaning
weaning_patients = df_weaning.dropna(subset=['endtime'])

# Initialize a list to store the final results
final_patients = []

# Iterate through weaning patients
for index, row in weaning_patients.iterrows():
    subject_id = row['subject_id']
    hadm_id = row['hadm_id']
    weaning_end_time = row['endtime']

    # Check if there was any intubation within 48 hours of weaning
    intubation_within_48 = df_intubation[(df_intubation['subject_id'] == subject_id) &
                        (df_intubation['hadm_id'] == hadm_id) &
                        (df_intubation['starttime'] <= weaning_end_time) &
                        (weaning_end_time - df_intubation['starttime'] <= pd.Timedelta('48 hours'))]

    # Check if there was any invasive ventilation within 48 hours of weaning
    ventilation_within_48 = df_invasive_ventilation[(df_invasive_ventilation['subject_id'] == subject_id) &
                              (df_invasive_ventilation['hadm_id'] == hadm_id) &
                              (df_invasive_ventilation['starttime'] <= weaning_end_time) &
                              (weaning_end_time - df_invasive_ventilation['starttime'] <= pd.Timedelta('48 hours'))]

    # Check if there was non-invasive ventilation within 48 hours of weaning
    non_invasive_ventilation = df_non_invasive_ventilation[(df_non_invasive_ventilation['subject_id'] == subject_id) &
                                  (df_non_invasive_ventilation['hadm_id'] == hadm_id) &
                                  (df_non_invasive_ventilation['starttime'] <= weaning_end_time) &
                                  (weaning_end_time - df_non_invasive_ventilation['starttime'] <= pd.Timedelta('48 hours'))]

    # Check if there was no intubation or ventilation within 48 hours of weaning, and non-invasive ventilation within 48 hours, then add the patient to the results
    if intubation_within_48.empty and ventilation_within_48.empty and (not non_invasive_ventilation.empty):
        final_patients.append(subject_id)

# Convert the list of patient IDs to a DataFrame
result_df = pd.DataFrame(final_patients, columns=['subject_id'])

# Print the result
print(result_df)

    subject_id
0     10150335
1     10321613
2     10438089
3     10887901
4     11321257
5     11576222
6     11648038
7     12365543
8     12365543
9     12765666
10    12953561
11    13333021
12    13788978
13    13880770
14    14188116
15    14479847
16    15790541
17    15897606
18    15983568
19    16099220
20    16455067
21    16544239
22    16702834
23    17275708
24    17302222
25    17475174
26    18119340
27    18119596
28    18266495
29    18308539
30    18489959
31    18571536
32    18784341
33    19362916
34    19386402
35    19589961
36    19851620


In [ ]:
# convert result_df to a list
patients_list = result_df['subject_id'].to_list()

#### no death within 48 h after weaning

In [ ]:
query = f"""
SELECT *
FROM `physionet-data.mimiciv_hosp.patients`
WHERE dod IS NOT NULL AND subject_id in {tuple(patients_list)}
"""

df_alive = run_query(query)
df_alive_patient_id = df_alive['subject_id'].to_list()
pd.DataFrame(df_alive_patient_id, columns=['subject_id'])

subject_id
0       10352416
1       16767959
2       15894414
3       12342976
4       10701372
...          ...
2938    12033229
2939    16643806
2940    11344751
2941    19149321
2942    17910586

[2943 rows x 1 columns]

### extract the patients' features after getting all the filtered paitents' id

read the filtered patients' id

In [62]:
patients_cohort = pd.read_csv('filtered_patients_dataset.csv', usecols=['subject_id'])
patients_cohort

subject_id
0       16467047
1       17087880
2       19556965
3       10222191
4       10252795
...          ...
5516    11102902
5517    14477446
5518    11230804
5519    15276613
5520    17411386

[5521 rows x 1 columns]

obtain the weaning start time point for all patients(assume invasive ventalition endtime as weaning start time)

In [63]:
query = f"""
SELECT char.subject_id, MIN(char.endtime) AS weaning_start_time, AVG(char.value) AS avg_ventilation_duration_in_minutes
FROM `physionet-data.mimiciv_icu.procedureevents` AS char
WHERE char.itemid = 225792
GROUP BY char.subject_id
ORDER BY char.subject_id
"""

df_first_invasive_ventilation = run_query(query)

# 将时长从分钟转换为天
df_first_invasive_ventilation['avg_ventilation_duration_in_days'] = df_first_invasive_ventilation['avg_ventilation_duration_in_minutes'] / 1440.0
df_first_invasive_ventilation = df_first_invasive_ventilation.drop(columns=['avg_ventilation_duration_in_minutes'])

# 合并到patients_cohort中
patients_cohort = patients_cohort.merge(df_first_invasive_ventilation, on="subject_id", how="left")
patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days
0       16467047 2189-10-20 11:15:00                          7.336111
1       17087880 2182-03-08 10:24:00                          0.975000
2       19556965 2168-04-30 18:30:00                          0.668750
3       10222191 2185-07-10 11:15:00                         22.254861
4       10252795 2149-11-16 15:30:00                          3.194444
...          ...                 ...                               ...
5516    11102902 2143-02-16 09:20:00                          0.304167
5517    14477446 2114-02-25 13:11:00                          6.563889
5518    11230804 2150-10-24 12:00:00                          0.277083
5519    15276613 2160-06-08 16:35:00                          0.156250
5520    17411386 2184-07-15 16:07:00                          0.171528

[5521 rows x 3 columns]

#### arterial blood gas

Arterial O2 pressure  220224 lowest

Arterial CO2 Pressure  220235 highest

Arterial Base Excess  224828 lowest

Inspired O2 Fraction  223835 higest

In [64]:
# 先获取patients_cohort中的所有subject_id
subject_ids = patients_cohort['subject_id'].tolist()

# 使用这些subject_id来筛选chartevents中的数据
query_o2 = f"""
SELECT subject_id, charttime, valuenum AS arterial_o2_pressure
FROM `physionet-data.mimiciv_icu.chartevents`
WHERE itemid = 220224 AND subject_id IN ({','.join(map(str, subject_ids))})
"""

df_o2 = run_query(query_o2)

# 使用weaning开始时间筛选24小时内的数据
df_o2 = df_o2.merge(patients_cohort[['subject_id', 'weaning_start_time']], on="subject_id", how="inner")
df_o2_filtered = df_o2[(df_o2['charttime'] > df_o2['weaning_start_time'] - pd.Timedelta(hours=24)) &
                       (df_o2['charttime'] <= df_o2['weaning_start_time'])]

# 筛选出每个病人在weaning开始前24小时内的最低arterial_o2_pressure值
df_o2_worst = df_o2_filtered.groupby('subject_id')['arterial_o2_pressure'].min().reset_index()

# 计算所有病人的arterial_o2_pressure的平均值
mean_o2 = df_o2_worst['arterial_o2_pressure'].mean()

# 对于没有arterial_o2_pressure数据的病人，使用平均值填充
missing_data = [{'subject_id': subject, 'arterial_o2_pressure': mean_o2} for subject in subject_ids if subject not in df_o2_worst['subject_id'].unique()]
df_o2_worst = pd.concat([df_o2_worst, pd.DataFrame(missing_data)], ignore_index=True)

# 合并df_o2_worst到patients_cohort
patients_cohort = patients_cohort.merge(df_o2_worst, on='subject_id', how='left')
patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  
0               100.000000  
1                20.000000  
2               114.000000  
3                36.000000  
4               109.946961  
...                    ...  
5516             94.000000  
5517            109.946961  
5518            315.000000  
5519             99.000000  
5520            108.000000  

[5521 rows x 4 columns]

In [74]:
def add_feature_to_cohort(patients_cohort, itemid, column_name, method='min'):
    subject_ids = patients_cohort['subject_id'].tolist()

    query = f"""
    SELECT subject_id, charttime, valuenum AS {column_name}
    FROM `physionet-data.mimiciv_icu.chartevents`
    WHERE itemid = {itemid} AND subject_id IN ({','.join(map(str, subject_ids))})
    """

    df_feature = run_query(query)

    # 使用weaning开始时间筛选24小时内的数据
    df_feature = df_feature.merge(patients_cohort[['subject_id', 'weaning_start_time']], on="subject_id", how="inner")
    df_feature_filtered = df_feature[(df_feature['charttime'] > df_feature['weaning_start_time'] - pd.Timedelta(hours=24)) &
                                     (df_feature['charttime'] <= df_feature['weaning_start_time'])]

    # 如果column_name不是'arterial_base_excess'，则删除值为0的记录
    if column_name != 'arterial_base_excess':
        df_feature_filtered = df_feature_filtered[df_feature_filtered[column_name] != 0]

    # 根据method选择获取每个病人在weaning开始前24小时内的最高或最低值
    if method == 'min':
        df_feature_worst = df_feature_filtered.groupby('subject_id')[column_name].min().reset_index()
    elif method == 'max':
        df_feature_worst = df_feature_filtered.groupby('subject_id')[column_name].max().reset_index()
    else:
        raise ValueError("Method should be either 'min' or 'max'.")

    # 计算所有病人的平均值
    mean_feature = df_feature_worst[column_name].mean()

    # 对于没有数据的病人，使用平均值填充
    missing_data = [{'subject_id': subject, column_name: mean_feature} for subject in subject_ids if subject not in df_feature_worst['subject_id'].unique()]
    df_feature_worst = pd.concat([df_feature_worst, pd.DataFrame(missing_data)], ignore_index=True)

    # 合并到patients_cohort
    return patients_cohort.merge(df_feature_worst, on='subject_id', how='left')

In [66]:
# 获取每个指标的最差值
patients_cohort = add_feature_to_cohort(patients_cohort, 220235, 'arterial_co2_pressure', method='max')  # 对CO2取最高值
patients_cohort = add_feature_to_cohort(patients_cohort, 224828, 'arterial_base_excess')  # 默认是获取最低值
patients_cohort = add_feature_to_cohort(patients_cohort, 223835, 'inspired_o2_fraction', method='max')  # 获取最高值

patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction  
0                     50.0  
1                    100.0  
2                     40.0  
3                     40.0  
4                     40.0  
...                    ...  
5516                  50.0  
5517                  35.0  
5518                  40.0  
5519                 100.0  
5520                 100.0  

[5521 rows x 7 columns]

#### full blood count

WBC 220546

platelet 227457

Hemoglobin	220228

In [67]:
# 使用定义的函数为patients_cohort添加新的特征
patients_cohort = add_feature_to_cohort(patients_cohort, 220546, "WBC", method='max')  # 对WBC取最高值
patients_cohort = add_feature_to_cohort(patients_cohort, 227457, "platelet")
patients_cohort = add_feature_to_cohort(patients_cohort, 220228, "Hemoglobin")
patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction   WBC  platelet  Hemoglobin  
0                     50.0  15.0     209.0         8.7  
1                    100.0   7.5      68.0         5.7  
2                     40.0  28.9     226.0         7.8  
3                     40.0  24.1     423.0         7.8  
4                     40.0   7.8     110.0        10.1  
...                    ...   ...       ...         ...  
5516                  50.0  16.9     281.0        10.8  
5517                  35.0   9.9     316.0         9.0  
5518                  40.0  16.3     176.0        12.4  
5519                 100.0  13.9     181.0        11.4  
5520                 100.0  23.3     139.0        10.6  

[5521 rows x 10 columns]

#### vital signs

Heart rate 220045 highest

Respiratory Rate	220210 highest

Arterial Blood Pressure mean	220052 lowest

Temperature Celsius	223762 highest

In [68]:
# 获取每个指标的最差值，并合并到patients_cohort
patients_cohort = add_feature_to_cohort(patients_cohort, 220045, 'Heart_rate', method='max')  # 获取最高值

patients_cohort = add_feature_to_cohort(patients_cohort, 220210, 'Respiratory_Rate', method='max')  # 获取最高值

patients_cohort = add_feature_to_cohort(patients_cohort, 220052, 'Arterial_BP_mean', method='min')  # 获取最低值

patients_cohort = add_feature_to_cohort(patients_cohort, 223762, 'Temperature_Celsius', method='max')  # 获取最高值

patients_cohort


subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction   WBC  platelet  Hemoglobin  Heart_rate  \
0                     50.0  15.0     209.0         8.7       111.0   
1                    100.0   7.5      68.0         5.7       132.0   
2                     40.0  28.9     226.0         7.8       126.0   
3                     40.0  24.1     423.0         7.8        87.0   
4                     40.0   7.8     110.0        10.1        77.0   
...                    ...   ...       ...         ...         ...   
5516                  50.0  16.9     281.0        10.8       111.0   
5517                  35.0   9.9     316.0         9.0        84.0   
5518                  40.0  16.3     176.0        12.4        74.0   
5519                 100.0  13.9     181.0        11.4       100.0   
5520                 100.0  23.3     139.0        10.6        90.0   

      Respiratory_Rate  Arterial_BP_mean  Temperature_Celsius  
0                 34.0         57.322394            37.541654  
1                 32.0         59.000000            37.541654  
2                 21.0         45.000000            37.541654  
3                 37.0         57.322394            37.541654  
4                 21.0         60.000000            37.541654  
...                ...               ...                  ...  
5516              19.0         54.000000            37.541654  
5517              27.0         61.000000            37.541654  
5518              25.0         78.000000            37.541654  
5519              21.0         64.000000            38.700000  
5520              40.0         41.000000            37.541654  

[5521 rows x 14 columns]

#### laboratory index

Creatinine (serum)	220615 highest

In [69]:
patients_cohort = add_feature_to_cohort(patients_cohort, 220615, 'Creatinine', method='max')  # 获取Creatinine的最高值

patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction   WBC  platelet  Hemoglobin  Heart_rate  \
0                     50.0  15.0     209.0         8.7       111.0   
1                    100.0   7.5      68.0         5.7       132.0   
2                     40.0  28.9     226.0         7.8       126.0   
3                     40.0  24.1     423.0         7.8        87.0   
4                     40.0   7.8     110.0        10.1        77.0   
...                    ...   ...       ...         ...         ...   
5516                  50.0  16.9     281.0        10.8       111.0   
5517                  35.0   9.9     316.0         9.0        84.0   
5518                  40.0  16.3     176.0        12.4        74.0   
5519                 100.0  13.9     181.0        11.4       100.0   
5520                 100.0  23.3     139.0        10.6        90.0   

      Respiratory_Rate  Arterial_BP_mean  Temperature_Celsius  Creatinine  
0                 34.0         57.322394            37.541654         1.0  
1                 32.0         59.000000            37.541654         2.3  
2                 21.0         45.000000            37.541654         2.5  
3                 37.0         57.322394            37.541654         1.8  
4                 21.0         60.000000            37.541654         1.1  
...                ...               ...                  ...         ...  
5516              19.0         54.000000            37.541654         0.5  
5517              27.0         61.000000            37.541654         0.6  
5518              25.0         78.000000            37.541654         0.8  
5519              21.0         64.000000            38.700000         0.7  
5520              40.0         41.000000            37.541654         0.9  

[5521 rows x 15 columns]

#### some other left features

age

gender

In [70]:
query_age_gender = """
SELECT subject_id, anchor_age AS age, gender
FROM `physionet-data.mimiciv_hosp.patients`
WHERE anchor_age >= 18
"""

age_gender_df = run_query(query_age_gender)

# 合并年龄和性别到patients_cohort
patients_cohort = patients_cohort.merge(age_gender_df, on='subject_id', how='left')
patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction   WBC  platelet  Hemoglobin  Heart_rate  \
0                     50.0  15.0     209.0         8.7       111.0   
1                    100.0   7.5      68.0         5.7       132.0   
2                     40.0  28.9     226.0         7.8       126.0   
3                     40.0  24.1     423.0         7.8        87.0   
4                     40.0   7.8     110.0        10.1        77.0   
...                    ...   ...       ...         ...         ...   
5516                  50.0  16.9     281.0        10.8       111.0   
5517                  35.0   9.9     316.0         9.0        84.0   
5518                  40.0  16.3     176.0        12.4        74.0   
5519                 100.0  13.9     181.0        11.4       100.0   
5520                 100.0  23.3     139.0        10.6        90.0   

      Respiratory_Rate  Arterial_BP_mean  Temperature_Celsius  Creatinine  \
0                 34.0         57.322394            37.541654         1.0   
1                 32.0         59.000000            37.541654         2.3   
2                 21.0         45.000000            37.541654         2.5   
3                 37.0         57.322394            37.541654         1.8   
4                 21.0         60.000000            37.541654         1.1   
...                ...               ...                  ...         ...   
5516              19.0         54.000000            37.541654         0.5   
5517              27.0         61.000000            37.541654         0.6   
5518              25.0         78.000000            37.541654         0.8   
5519              21.0         64.000000            38.700000         0.7   
5520              40.0         41.000000            37.541654         0.9   

      age gender  
0      74      F  
1      59      F  
2      69      F  
3      77      M  
4      72      M  
...   ...    ...  
5516   69      F  
5517   88      F  
5518   65      M  
5519   54      M  
5520   69      M  

[5521 rows x 17 columns]

In [71]:
dod_gender = """
SELECT subject_id, dod
FROM `physionet-data.mimiciv_hosp.patients`
"""

dod_df = run_query(dod_gender)

# 合并年龄和性别到patients_cohort
patients_cohort = patients_cohort.merge(dod_df, on='subject_id', how='left')
patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction   WBC  platelet  Hemoglobin  Heart_rate  \
0                     50.0  15.0     209.0         8.7       111.0   
1                    100.0   7.5      68.0         5.7       132.0   
2                     40.0  28.9     226.0         7.8       126.0   
3                     40.0  24.1     423.0         7.8        87.0   
4                     40.0   7.8     110.0        10.1        77.0   
...                    ...   ...       ...         ...         ...   
5516                  50.0  16.9     281.0        10.8       111.0   
5517                  35.0   9.9     316.0         9.0        84.0   
5518                  40.0  16.3     176.0        12.4        74.0   
5519                 100.0  13.9     181.0        11.4       100.0   
5520                 100.0  23.3     139.0        10.6        90.0   

      Respiratory_Rate  Arterial_BP_mean  Temperature_Celsius  Creatinine  \
0                 34.0         57.322394            37.541654         1.0   
1                 32.0         59.000000            37.541654         2.3   
2                 21.0         45.000000            37.541654         2.5   
3                 37.0         57.322394            37.541654         1.8   
4                 21.0         60.000000            37.541654         1.1   
...                ...               ...                  ...         ...   
5516              19.0         54.000000            37.541654         0.5   
5517              27.0         61.000000            37.541654         0.6   
5518              25.0         78.000000            37.541654         0.8   
5519              21.0         64.000000            38.700000         0.7   
5520              40.0         41.000000            37.541654         0.9   

      age gender         dod  
0      74      F  2189-11-21  
1      59      F  2182-03-08  
2      69      F  2168-04-30  
3      77      M  2185-09-01  
4      72      M  2149-11-19  
...   ...    ...         ...  
5516   69      F         NaT  
5517   88      F  2114-02-27  
5518   65      M         NaT  
5519   54      M         NaT  
5520   69      M         NaT  

[5521 rows x 18 columns]

#### classify WF and WS

In [79]:
def classify_weaning_success(patients_cohort):
    subject_ids = patients_cohort['subject_id'].tolist()

    # 获取非侵入性通气的记录
    non_invasive_vent_query = f"""
    SELECT subject_id, starttime, value
    FROM `physionet-data.mimiciv_icu.procedureevents`
    WHERE itemid = 225794 AND subject_id IN ({','.join(map(str, subject_ids))})
    """
    non_invasive_vent_df = run_query(non_invasive_vent_query)

    # 获取侵入性通气的记录
    invasive_vent_query = f"""
    SELECT subject_id, starttime
    FROM `physionet-data.mimiciv_icu.procedureevents`
    WHERE itemid = 225792 AND subject_id IN ({','.join(map(str, subject_ids))})
    """
    invasive_vent_df = run_query(invasive_vent_query)

    # 根据条件分类
    def classify_row(row):
        weaning_time = row['weaning_start_time']
        subject = row['subject_id']

        # 检查条件 (c)
        non_inv_vent_records = non_invasive_vent_df[non_invasive_vent_df['subject_id'] == subject]
        within_48_hours_records = non_inv_vent_records[(non_inv_vent_records['starttime'] > weaning_time) &
                                                       (non_inv_vent_records['starttime'] <= weaning_time + pd.Timedelta(hours=48))]
        total_duration = within_48_hours_records['value'].sum()
        if total_duration >= 48*60:  # 将时间从分钟转换为48小时
            return "WF"

        # 检查条件 (a)
        inv_vent_records = invasive_vent_df[invasive_vent_df['subject_id'] == subject]
        if any((inv_vent_records['starttime'] > weaning_time) &
               (inv_vent_records['starttime'] <= weaning_time + pd.Timedelta(hours=48))):
            return "WF"

        # 检查条件 (b)
        death_date = row['dod']
        if not pd.isnull(death_date) and pd.Timestamp(death_date) <= weaning_time + pd.Timedelta(hours=48):
            return "WF"

        return "WS"

    patients_cohort['weaning_class'] = patients_cohort.apply(classify_row, axis=1)
    return patients_cohort

patients_cohort = classify_weaning_success(patients_cohort)
patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction   WBC  platelet  Hemoglobin  ...  \
0                     50.0  15.0     209.0         8.7  ...   
1                    100.0   7.5      68.0         5.7  ...   
2                     40.0  28.9     226.0         7.8  ...   
3                     40.0  24.1     423.0         7.8  ...   
4                     40.0   7.8     110.0        10.1  ...   
...                    ...   ...       ...         ...  ...   
5516                  50.0  16.9     281.0        10.8  ...   
5517                  35.0   9.9     316.0         9.0  ...   
5518                  40.0  16.3     176.0        12.4  ...   
5519                 100.0  13.9     181.0        11.4  ...   
5520                 100.0  23.3     139.0        10.6  ...   

      Temperature_Celsius  Creatinine  age  gender         dod  PEEP_Level  \
0               37.541654         1.0   74       F  2189-11-21    7.456549   
1               37.541654         2.3   59       F  2182-03-08    6.000000   
2               37.541654         2.5   69       F  2168-04-30    7.000000   
3               37.541654         1.8   77       M  2185-09-01    7.456549   
4               37.541654         1.1   72       M  2149-11-19    5.000000   
...                   ...         ...  ...     ...         ...         ...   
5516            37.541654         0.5   69       F         NaT    7.456549   
5517            37.541654         0.6   88       F  2114-02-27    7.456549   
5518            37.541654         0.8   65       M         NaT    5.000000   
5519            38.700000         0.7   54       M         NaT    7.456549   
5520            37.541654         0.9   69       M         NaT    7.456549   

      SpO2 Anion_gap  Urine_output  weaning_class  
0     85.0      15.0     575.51594             WS  
1     85.0      32.0     575.51594             WF  
2     85.0      48.0     575.51594             WF  
3     85.0      16.0     575.51594             WF  
4     85.0      12.0     575.51594             WS  
...    ...       ...           ...            ...  
5516  85.0      12.0    1600.00000             WS  
5517  88.0      10.0     575.51594             WF  
5518  88.0      14.0     150.00000             WS  
55

#### additional
226566     Urine and GU Irrigant Out minimum

224700       Total PEEP Level highest

226253       SpO2 Desat Limit minimum

Anion gap	227073 highest

In [76]:
def add_feature_to_cohort(patients_cohort, itemid, column_name, table_name='chartevents', method='min'):
    subject_ids = patients_cohort['subject_id'].tolist()

    query = f"""
    SELECT subject_id, charttime, value AS {column_name}
    FROM `physionet-data.mimiciv_icu.{table_name}`
    WHERE itemid = {itemid} AND subject_id IN ({','.join(map(str, subject_ids))})
    """

    df_feature = run_query(query)

    # 使用weaning开始时间筛选24小时内的数据
    df_feature = df_feature.merge(patients_cohort[['subject_id', 'weaning_start_time']], on="subject_id", how="inner")
    df_feature_filtered = df_feature[(df_feature['charttime'] > df_feature['weaning_start_time'] - pd.Timedelta(hours=24)) &
                                     (df_feature['charttime'] <= df_feature['weaning_start_time'])]

    # 如果column_name不是'arterial_base_excess'，则删除值为0的记录
    if column_name != 'arterial_base_excess':
        df_feature_filtered = df_feature_filtered[df_feature_filtered[column_name] != 0]

    # 根据method选择获取每个病人在weaning开始前24小时内的最高或最低值
    if method == 'min':
        df_feature_worst = df_feature_filtered.groupby('subject_id')[column_name].min().reset_index()
    elif method == 'max':
        df_feature_worst = df_feature_filtered.groupby('subject_id')[column_name].max().reset_index()
    else:
        raise ValueError("Method should be either 'min' or 'max'.")

    # 计算所有病人的平均值
    mean_feature = df_feature_worst[column_name].mean()

    # 对于没有数据的病人，使用平均值填充
    missing_data = [{'subject_id': subject, column_name: mean_feature} for subject in subject_ids if subject not in df_feature_worst['subject_id'].unique()]
    df_feature_worst = pd.concat([df_feature_worst, pd.DataFrame(missing_data)], ignore_index=True)

    # 合并到patients_cohort
    return patients_cohort.merge(df_feature_worst, on='subject_id', how='left')


In [77]:
patients_cohort = add_feature_to_cohort(patients_cohort, 226627, 'Urine_output',table_name = 'outputevents', method='min')
#patients_cohort = add_feature_to_cohort(patients_cohort, 224700, 'PEEP_Level', method='max')
#patients_cohort = add_feature_to_cohort(patients_cohort, 226253, 'SpO2', method='min')
#patients_cohort = add_feature_to_cohort(patients_cohort, 227073, 'Anion_gap', method='max')
patients_cohort

subject_id  weaning_start_time  avg_ventilation_duration_in_days  \
0       16467047 2189-10-20 11:15:00                          7.336111   
1       17087880 2182-03-08 10:24:00                          0.975000   
2       19556965 2168-04-30 18:30:00                          0.668750   
3       10222191 2185-07-10 11:15:00                         22.254861   
4       10252795 2149-11-16 15:30:00                          3.194444   
...          ...                 ...                               ...   
5516    11102902 2143-02-16 09:20:00                          0.304167   
5517    14477446 2114-02-25 13:11:00                          6.563889   
5518    11230804 2150-10-24 12:00:00                          0.277083   
5519    15276613 2160-06-08 16:35:00                          0.156250   
5520    17411386 2184-07-15 16:07:00                          0.171528   

      arterial_o2_pressure  arterial_co2_pressure  arterial_base_excess  \
0               100.000000              51.000000             -1.000000   
1                20.000000              75.000000            -18.000000   
2               114.000000              37.000000            -20.000000   
3                36.000000              49.000000              0.000000   
4               109.946961             475.305562             -2.995491   
...                    ...                    ...                   ...   
5516             94.000000              48.000000              0.000000   
5517            109.946961             475.305562             -2.995491   
5518            315.000000              36.000000              0.000000   
5519             99.000000              57.000000             -4.000000   
5520            108.000000              52.000000             -5.000000   

      inspired_o2_fraction   WBC  platelet  Hemoglobin  ...  Arterial_BP_mean  \
0                     50.0  15.0     209.0         8.7  ...         57.322394   
1                    100.0   7.5      68.0         5.7  ...         59.000000   
2                     40.0  28.9     226.0         7.8  ...         45.000000   
3                     40.0  24.1     423.0         7.8  ...         57.322394   
4                     40.0   7.8     110.0        10.1  ...         60.000000   
...                    ...   ...       ...         ...  ...               ...   
5516                  50.0  16.9     281.0        10.8  ...         54.000000   
5517                  35.0   9.9     316.0         9.0  ...         61.000000   
5518                  40.0  16.3     176.0        12.4  ...         78.000000   
5519                 100.0  13.9     181.0        11.4  ...         64.000000   
5520                 100.0  23.3     139.0        10.6  ...         41.000000   

      Temperature_Celsius  Creatinine  age  gender         dod PEEP_Level  \
0               37.541654         1.0   74       F  2189-11-21   7.456549   
1               37.541654         2.3   59       F  2182-03-08   6.000000   
2               37.541654         2.5   69       F  2168-04-30   7.000000   
3               37.541654         1.8   77       M  2185-09-01   7.456549   
4               37.541654         1.1   72       M  2149-11-19   5.000000   
...                   ...         ...  ...     ...         ...        ...   
5516            37.541654         0.5   69       F         NaT   7.456549   
5517            37.541654         0.6   88       F  2114-02-27   7.456549   
5518            37.541654         0.8   65       M         NaT   5.000000   
5519            38.700000         0.7   54       M         NaT   7.456549   
5520            37.541654         0.9   69       M         NaT   7.456549   

      SpO2  Anion_gap  Urine_output  
0     85.0       15.0     575.51594  
1     85.0       32.0     575.51594  
2     85.0       48.0     575.51594  
3     85.0       16.0     575.51594  
4     85.0       12.0     575.51594  
...    ...        ...           ...  
5516  85.0       12.0    1600.00000  
5517  88.0    

In [80]:
patients_cohort.to_csv('filtered_patients_dataset_v2.csv', index=False)